src/coreclr/jit/lower.cpp

```diff
@@ -6776,10 +6776,10 @@ bool Lowering::CheckMultiRegLclVar(GenTreeLclVar* lclNode, const ReturnTypeDesc*
         }
     }
 #ifdef TARGET_XARCH
-    // For local stores on XARCH we only handle mismatched src/dest register count for
-    // calls of SIMD type. If the source was another lclVar similarly promoted, we would
+    // For local stores on XARCH we can't handle another lclVar source.
+    // If the source was another lclVar similarly promoted, we would
     // have broken it into multiple stores.
-    if (lclNode->OperIs(GT_STORE_LCL_VAR) && !lclNode->gtGetOp1()->OperIs(GT_CALL))
+    if (lclNode->OperIs(GT_STORE_LCL_VAR) && lclNode->gtGetOp1()->OperIs(GT_LCL_VAR))
```

Comment:
Did you try to find the root cause of this?

Reply:
I didn't have time for this. I brought it from another PR mentioned earlier and it just works. It's better to ask the original author.

Reply:
@CarolEidt can you help explaining this? Not doing this isn't causing failures now, but the code looks necessary.

Reply:
@huoyaoyuan - Carol doesn't work on this code base anymore. As the comment suggests, " If the source was another lclVar similarly promoted, we would have broken it into multiple stores.". So, I think you should revert this change, given that it works now.

Reply:
It's better to provide an example involving this path so that I can learn from debugging. Currently, neither simple example nor the original failing `TryFormatDateTimeO` is triggering the caller method `LowerStoreLocCommon`.

Reply:
It's not easy to debug this. It seems that `lvaEnregMultiRegVars` is only true under Release CLR.

Should it be changed to another check for `GT_HWINTRINSIC`, aside of `GT_CALL`?

Reply:
> It seems that `lvaEnregMultiRegVars` is only true under Release CLR.

(Are you running with optimizations on, aka TC off?)

> Should it be changed to another check for `GT_HWINTRINSIC`, aside of `GT_CALL`?

It's not clear that would be correct. Consider: `(uint, uint) a = DivRem(...)`, with the tuple on the LHS not promoted. I suspect it won't work right currently, with `a` enregistered as `long`. It is more of a problem with the check itself of course, as it only DNERs promoted locals.

Really, it seems the logic should be more like this:
```cs
if (comp->lvaEnregMultiRegVars && varDsc->lvPromoted) { ... }
else
{
    canEnreg = varTypeIsSIMD(lclNode);
#ifdef TARGET_XARCH
    // The current clause with the call.
#endif
}

if (canEnreg)
{
    lclNode->SetMultiReg();
}
else
{
    lclNode->ClearMultiReg();
    if (!varDsc->lvDoNotEnregister)
    {
        comp->lvaSetVarDoNotEnregister(lclNode->GetLclNum() DEBUGARG(DoNotEnregisterReason::BlockOp));
    }
}
```

Reply:
The change is involving more and I have really no confidence to not break others. Who is the one with expertise in this?

Reply:
I will spend some time to understand in more depth.

Reply:
Reverting seems not valid - it's causing various different failures in CI.
But I have no idea why these cases are failing.

Reply:
Yes, that was just a means to try which nodes are caught if `op1 != GT_CALL`. So the reason it is failing (as expected) is because the tree now contains the `op1 == GT_HWINTRINSIC`.

```
                                                            ┌──▌  t10    int    
                                                            ├──▌  t13    int    
                                                            ├──▌  t58    int    
N007 (  8,  9) [000014] -----------                   t14 = ▌  HWINTRINSIC struct int DivRem $140
                                                            ┌──▌  t14    struct 
N009 ( 18, 16) [000017] MA---------                         ▌  STORE_LCL_VAR struct<System.ValueTuple`2[System.Int32, System.Int32], 8>(P) V05 tmp3         
                                                            ▌    int    V05.Item1 (offs=0x00) -> V13 tmp11        
                                                            ▌    int    V05.Item2 (offs=0x04) -> V14 tmp12   
```

We do want to enregister in such cases and not doing it (retaining `op1 != GT_CALL`) spills the result on stack.

https://www.diffchecker.com/TUY5sGjV

Reply:
> Isn't this the same code on xarch?

Yes and no. My suggestion includes the (significant) addition that we will DNER _all_ locals, not just promoted ones. It is to prevent the case I described above from arising: `long = multi-reg-node` and other assigns codegen doesn't support. Alternatively, we could of course support them in codegen, but that seems out of scope for this change.

> Also, for the below code, do you mean to say use the current check of independent promotion too?

Yes, the independent promotion check (and the field count check as well) must stay. In fact, I think they should be expanded. Consider `struct { promoted field<double> } = HWI(int, int)` and other mismatched cases. Right now they would pass, but we need to DNER them, by passing the register count to the method instead of the return type descriptor.

Reply:
There's now new failures and I don't have enough time to investigate in more depth.

Reply:
I will take a look hopefully next week.

Reply:
I updated the code as per @SingleAccretion suggestion.